test/integration/acme/orders_controller_test.go

```diff
@@ -0,0 +1,280 @@
+
+	// Create an Order CR.
+	order := gen.Order(testName,
+		gen.SetOrderIssuer(cmmeta.ObjectReference{
+			Name: testName,
+		}),
+		gen.SetOrderNamespace(testName),
+		gen.SetOrderCsr([]byte(testName)),
+		gen.SetOrderDNSNames(testName))
+
+	order, err = cmCl.AcmeV1().Orders(testName).Create(ctx, order, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the Challenge to be created.
+	var chal *cmacme.Challenge
+	err = wait.Poll(time.Millisecond*100, time.Minute, func() (done bool, err error) {
+		chals, err := cmCl.AcmeV1().Challenges(testName).List(ctx, metav1.ListOptions{})
+		l := len(chals.Items)
+		// Challenge has not been created yet
+		if l == 0 {
+			return false, nil
+		}
+		// this should never happen
+		if l > 1 {
+			return false, fmt.Errorf("expected maximum 1 challenge, got %d", l)
+		}
+		// Check that the Challenge is owned by our Order.
+		chal = &chals.Items[0]
+		if !metav1.IsControlledBy(chal, order) {
+			return false, fmt.Errorf("found an unexpected Challenge resource: %v", chal.Name)
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// 2. Test that in an edge case, where an ACME server is misbehaving and
+	// despite Challenges being valid, the ACME order status is 'pending', we
+	// re-queue the Order so that when the ACME order does become 'ready', we
+	// finalize our Order and, in the success scenario, it eventually becomes
+	// valid.
+	// https://github.com/jetstack/cert-manager/issues/2868
+
+	// Set the Challenge state to valid- the status of the ACME order remains 'pending'.
+	chal = chal.DeepCopy()
+	chal.Status.State = cmacme.Valid
+	_, err = cmCl.AcmeV1().Challenges(testName).UpdateStatus(ctx, chal, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Sit here for a little bit checking that the Order status remains pending
+	// and also to verify that this test works.
+	var pendingOrder *cmacme.Order
+	err = wait.Poll(time.Millisecond*200, time.Second*5, func() (bool, error) {
+		pendingOrder, err = cmCl.AcmeV1().Orders(testName).Get(ctx, testName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if pendingOrder.Status.State != cmacme.Pending {
+			return true, nil
+		}
+		return false, nil
+	})
+	switch {
+	case err == nil:
+		t.Fatalf("Expected Order to have pending status instead got: %v", pendingOrder.Status.State)
+	case err == wait.ErrWaitTimeout:
+		// 'happy case' - Order remained pending
+	default:
+		t.Fatal(err)
+	}
+
+	// Set status of the ACME order to 'ready.
+	acmeOrder.Status = acmeapi.StatusReady
+
+	// Wait for the status of the Order to become Valid.
+	err = wait.Poll(time.Millisecond*100, time.Second*20, func() (bool, error) {
+		o, err := cmCl.AcmeV1().Orders(testName).Get(ctx, testName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// not valid yet
+		if o.Status.State != cmacme.Valid {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
```

Comment:
Perhaps this should patch the RequeuePeriod to something smaller and then wait some multiple of the RequeuePeriod here....rather than `time.Second *20`